include/arch/arm/arch/64/mode/object/structures.bf

```diff
@@ -126,8 +126,8 @@ block sid_control_cap {
 
 block sid_cap {
 
-    padding              52
-    field capSID         12
+    padding              48
+    field capSID         16
```

Comment:
Can you please elaborate on why this is necessary and what it fixes?

Reply:
I guess that is probably the stream ID bit size is increased on the hardware so that more devices can be supported.

Reply:
I agree, probably, but the PR should really explain that not just change numbers ;-)

My concern is that if this is different between different boards, we either need to find a maximum that suits all and then check that we're not using values outside the range at runtime, or we need to use config information here and try to do this at compile time, adjusting padding and other fields as long as there is space.

I suspect we should the a static maximum over all boards for compile time and add the runtime check against board config in the decode functions where this is used.

Reply:
And I now found the commit message that actually says this, sorry, should have seen that earlier (although it would have helped to put that info into the PR description -- review time is a scarce resource).

So this is now a conflict between what the TX2 board provides and what the SMMUv2 spec seems to say. Do we still take the maximum?

Reply:
> I agree, probably, but the PR should really explain that not just change numbers ;-)
> 
> My concern is that if this is different between different boards, we either need to find a maximum that suits all and then check that we're not using values outside the range at runtime, or we need to use config information here and try to do this at compile time, adjusting padding and other fields as long as there is space.
> 
> I suspect we should the a static maximum over all boards for compile time and add the runtime check against board config in the decode functions where this is used.

Yes, I think this is a good approach: a static max number and a runtime probe to get the actual number. I hope the static max numbers are not too big, so we do not waste too much resources.

Reply:
My apologies, I will elaborate more in the PR description in the future. I appreciate the feedback. Yes, the goal of this change was to use a static maximum taken from the SMMUv2 spec so that it can more broadly support additional boards in the future. In this specific use case, the zynqmp utilizes 15 bits (MAX_SID 32767) as opposed the 7 implemented by the TX2.

Reply:
There was an issue associated with extra changes that would be required when supporting 16bit StreamIDs: https://sel4.atlassian.net/browse/SELFOUR-2800

Reply:
Hm, with 2^15 ctes that'd be 2^20 bytes, i.e. 1MB. That is a lot, but supposedly if you have a board with such a large space you'd expect it to have a lot of memory as well. Would it really be that bad to allocate that much? Doing a two-level table as described in the issue would mean that the user has to dynamically manage it (like with ASID pools). It's doable, but it would be a fairly large API change and additional mechanism.

Reply:
Well the difference is that TX2 allocates SIDs to devices in order.  ZCU102 is a lot more sparse as they use some part of the ID space to encode a tag that has meaning to the hardware (I think it affects which hardware TLB gets used).  So it may be possible to make some platform specific translation that compresses the SID back to a continuous range.

Reply:
Bumping the conversation. Not sure how to proceed here. Do we need to have it dynamically managed to get this through? Are there any repercussions for merging this as is? Sounds like a significant amount of work to get dynamic management working. This change at least gets it working on the zynqmp platforms in the mean time.

Reply:
@Apave24, are you able to explain a bit more how the SID encodings work for the zynqmp boards? According to the documentaion: "The SMMU uses a 15-bit stream ID to perform address translations. This information is part of a transaction and indicates which master originated the request. Bits [9:0] of the stream ID are the master ID defined in Table 16-13. Bits [14:10] are the TBU number the master transaction passes through. For example, for the GEM0 PS master, its master ID is 10'h074 and its TBU number is 5'h02. Concatenating these fields gives GEM0 a stream ID of
15'h0874."

If you were to enumerate all of the possible SID values that a zynqmp board has is it 2^15 or closer to the number of master IDs (2^9)?

Reply:
The total number of possible SID values is probably closer to the number of master IDs (2^10), however, the zynqmp uses 15 bits for the IDs.

Although should this discussion be holding up this PR? These commits just allow for a potential ID value up to the maximum value per the SMMUv2 spec and fixes a build time error. The amount of memory allocated is based on the SMMU_MAX_SID which is set on a per platform basis which is added for the zynqmp in PR #761 . So maybe changes in that PR need to be reworked a bit and this one is good to go?

Reply:
Reduced the cnode size in PR #761 commit [3841eca](https://github.com/seL4/seL4/pull/761/commits/3841eca2ec2d6e2b9c31ba3e2410cec1c9d29b3e)